### DISCUSSION TITLE
Matlab: MATLAB toolbox functions shadow DIPimage functions

### DISCUSSION THREAD

**Discussion by AndreZeug** [DIPlib/diplib#189] [General]
In some systems (e.g. where all MATLAB Toolbox functions got installed by the admin) MATLAB could shadow DIPimage functions and the user has no direct possibilities to uninstall toolboxes causing trouble (in case they are not needed).
```
>> addpath('/project/cnp/DIPlib/share/DIPimage','-begin')
>> which -all threshold
/hpc/leinesw/modules/software/noarch/MATLAB/2023a/toolbox/econ/econ/@threshold/threshold.m  % threshold constructor
/hpc/project/cnp/DIPlib/share/DIPimage/threshold.m                                          % Shadowed
```
For some reason the Tweak discussed here [https://diplib.org/diplib-docs/sec_dum_faq.html#sec_dum_faq_usage_select](https://diplib.org/diplib-docs/sec_dum_faq.html#sec_dum_faq_usage_select) does not work. 

An alternative to the "changing folder approach" each time before the function is called is removing the path to Toolbox functions which are not needed.
This can be achieved by adding, e.g., 
```
rmpath([matlabroot, '/toolbox/econ/econ:'])
```
to the `startup.m` .
```
>> addpath('/project/cnp/DIPlib/share/DIPimage','-begin')
>> rmpath([matlabroot, '/toolbox/econ/econ:'])
>> which -all threshold
/hpc/project/cnp/DIPlib/share/DIPimage/threshold.m
```
This resolved my issue. 

To address the issue of **Provide User control of MATLAB Toolboxes, which where installed by the admin** I got in contact with MATLAB Technical Support and requested:  

> Maybe Mathworks could provide an GUI for users, similar to the installation tool for admins, to select / unselect toolboxes according to the users' needs which then amends the MATLAB path. 

MATLAB Technical Support replied 

> Unfortunately, there isn't a MATLAB official way of easily stopping toolboxes from being loaded up on startup. The only way would be to modify the MATLAB path and remove them, so MATLAB doesn't see them, then add them back in when you need them.
> The command `pathtool` will open up which MATLAB paths you can see, and there may be a way to write a script to remove/add toolbox paths, but this would be unsupported by MATLAB support. 
> However, I think your request for such a GUI for users would be reasonable. I will create an enhancement for this to bring it to the attention of our developers and link your case to our internal documentation. 

Since such tool is not available (yet), MATLAB Technical Support recommended the following alternative information:

> "threshold" in the Econometrics Toolbox is not a built-in function, it is implemented in M-code.
> You can take a look at the "More About" section at the following link: https://www.mathworks.com/help/matlab/ref/builtin.html#bthvewb-5
> "A built-in function is part of the MATLAB executable. MATLAB does not implement these functions in the MATLAB language."
> Unfortunately, we do not have any clean way of calling a shadowed non-builtin function.
> To work around this issue, you can use "which -all", as you did already, to find the shadowed function and create a function handle to it and use the function handle to call the shadowed function.
> Here is a MATLAB answer that may help.
> https://www.mathworks.com/matlabcentral/answers/149059-how-do-i-invoke-a-shadowed-core-matlab-function-not-built-in-from-an-overloaded-function-of-same
> However, please be aware that the behavior here is subject to change in future releases. You can find more information here: https://www.mathworks.com/help/releases/R2024b/matlab/matlab_oop/organizing-classes-in-folders.html#brfynrp-5
> So actually, I would recommend making sure that the function you have in your project has a unique name, to avoid unforeseen behavior in the future.

I hope this information help.

Please see also **[Matlab: DIPimage shadows Matlab toolbox functions](https://github.com/DIPlib/diplib/discussions/174)**

**Comment by crisluengo** [comment]
I didn't realize that class constructors have precedence over regular functions. According to [this doc page](https://www.mathworks.com/help/matlab/matlab_prog/function-precedence-order.html), this is indeed the case:

> [...]
> 8\. Class constructors in @-folders
> [...]
> 11\. Functions elsewhere on the path, in order of appearance

So, just putting the DIPimage directory first in the path is not enough to always select DIPimage functions, if there exists a class anywhere in the path that has the same name of the function.

That is... unfortunate.

That makes it even harder for 3rd party toolbox authors to maintain their toolboxes. They basically expect us to rename our functions every time they add a class to any of the official toolboxes?

I think the best you can do is edit your path manually once (using `pathtool`), then in your startup folder type `savepath pathdef.m`. This will create a `pathdef.m` file that shadows the system-wide one (the one that your sysadmin controls) and will be loaded every time you start MATLAB.
Do Note that the base MATLAB has a lot of paths all by itself, so don't just remove everything you don't recognize.

I agree that a GUI that allows you to select which toolboxes you want enabled by default would be very useful for larger institutions that subscribe to many toolboxes that you personally don't have a use for. I hope they build something like that in.
